### PR TITLE
DeepZ Certification

### PR DESCRIPTION
**Pull request by GiulioZizzo**
# Description

We include a certification estimator based on the zonotope domain. We use the DeepZ ReLU relaxation proposed in: Singh, Gagandeep, et al. "Fast and effective robustness certification." Advances in neural information processing systems 31 (2018).

Fixes # (issue) NA

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Tests are located in tests/estimators/certification/test_deepz.py . These tests check the bounds produced by the code against known ground truths as well as classifier performance.

**Test Configuration**:
- OS: Ubuntu 18
- Python version: 3.7.10
- ART version or commit number: 1.9
- TensorFlow: 2.3.1 / Keras: 2.4.3 / PyTorch: 1.8.1 / MXNet version: NA

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/certification/deep_z/deep_z.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        bias = torch.unsqueeze(self.bias, dim=-1)
+        bias = torch.unsqueeze(bias, dim=-1)
+        x[0] = x[0] + bias
+        return x
+
+
+class ZonoReLU(torch.nn.Module, ZonoBounds):
+    """
+    Implements "DeepZ" for relu.
+    Paper:  https://papers.nips.cc/paper/2018/file/f2f446980d8e971ef3da97af089481c3-Paper.pdf
```
Comment (beat-buesser) [inline art/estimators/certification/deep_z/deep_z.py:253]:
```suggestion
    Implements "DeepZ" for relu.
    
    | Paper link:  https://papers.nips.cc/paper/2018/file/f2f446980d8e971ef3da97af089481c3-Paper.pdf
```

File: art/estimators/certification/deep_z/deep_z.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+        param x: input zonotope to the dense layer.
+        return x: zonotope after being pushed through the dense layer.
+        """
+        return self.zonotope_relu(x)
+
+    def zonotope_relu(self, x: "torch.Tensor") -> "torch.Tensor":
+        """
+        Implements "DeepZ" for relu.
+        :param x: input zonotope
+        :return x: zonotope after application of the relu. May have grown in dimension if crossing relus occur.
```
Comment (beat-buesser) [inline art/estimators/certification/deep_z/deep_z.py:275]:
```suggestion
        Implements "DeepZ" for relu.

        :param x: input zonotope
        :return x: zonotope after application of the relu. May have grown in dimension if crossing relus occur.
```

File: art/estimators/certification/deep_z/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+if TYPE_CHECKING:
+    from art.utils import CLIP_VALUES_TYPE, PREPROCESSING_TYPE
+    from art.defences.preprocessor import Preprocessor
+    from art.defences.postprocessor import Postprocessor
+
+
+class PytorchDeepZ(PyTorchClassifier, ZonoBounds):
+    """
+    Implementation of DeepZ (https://papers.nips.cc/paper/2018/file/f2f446980d8e971ef3da97af089481c3-Paper.pdf)
```
Comment (beat-buesser) [inline art/estimators/certification/deep_z/pytorch.py:42]:
```suggestion
    Implementation of DeepZ
```

File: art/estimators/certification/deep_z/pytorch.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    estimator_params = PyTorchClassifier.estimator_params
+
+    def __init__(
+        self,
+        model: "torch.nn.Module",
+        loss: "torch.nn.modules.loss._Loss",
+        input_shape: Tuple[int, ...],
+        nb_classes: int,
+        optimizer: Optional["torch.optim.Optimizer"] = None,  # type: ignore
```
Comment (beat-buesser) [inline art/estimators/certification/deep_z/pytorch.py:58]: Why is the type ignore applied?

File: tests/estimators/certification/test_deepz.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+import json
+import os
+from art.utils import load_dataset
+from art.estimators.certification.deep_z import PytorchDeepZ
+import pytest
+import numpy as np
+from torch import nn
+import torch
+from tests.utils import ARTTestException
+from tests.utils import get_image_classifier_pt, get_cifar10_image_classifier_pt
```
Comment (beat-buesser) [inline tests/estimators/certification/test_deepz.py:28]: Please order imports according to PEP8.

**Comment by beat-buesser** [comment]
Could you please update all docstrings to follow this pattern:
```python
    """
    Convert a 2D-array of floating point labels to binary class matrix.

    :param labels: A 2D-array of floating point labels of shape `(nb_samples, nb_classes)`
    :return: A binary matrix representation of `labels` in the shape `(nb_samples, nb_classes)`
    """
```
including and empty line after the general description, a param line for each argument and a return line for the returned values?

This is important to enable the correct automatic generation of the documentation.

File: run_tests.sh [hunk 6]
```diff
@@ -61,6 +61,9 @@ then
     pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/estimators/classification/test_deeplearning_specific.py --framework=$framework --durations=0
     if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed estimators/classification tests for framework $framework"; fi
 
+    pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/estimators/certification/test_deepz.py --framework=$framework --durations=0
+    if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed estimators/certification tests for framework $framework"; fi
+
```
Comment (beat-buesser) [inline run_tests.sh:66]:
```suggestion
    pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/estimators/certification/ --framework=$framework --durations=0
    if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed estimators/certification tests for framework $framework"; fi

```

File: art/estimators/certification/deep_z/pytorch.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+                zono_conv = ZonoConv(
+                    in_channels=module.in_channels,
+                    out_channels=module.out_channels,
+                    kernel_size=module.kernel_size,  # type: ignore
+                    stride=module.stride,  # type: ignore
+                    dilation=module.dilation,  # type: ignore
+                    padding=module.padding,  # type: ignore
+                )
+                zono_conv.conv.weight.data = module.weight.data.to(self.device)
+                zono_conv.bias.data = module.bias.data.to(self.device)  # type: ignore
```
Comment (beat-buesser) [inline art/estimators/certification/deep_z/pytorch.py:141]: Why are the type ignores required?

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @GiulioZizzo Thank you very much for implementing DeepZ certification in PyTorch!

I have made a few review comments, let me know what you think!

Do you think it could be worth adding a notebook demonstrating the application of DeepZ?

File: tests/estimators/certification/test_deepz.py (hunk 5 above)
Comment (GiulioZizzo) [reply tests/estimators/certification/test_deepz.py]: import order changed appropriately

File: art/estimators/certification/deep_z/pytorch.py (hunk 4 above)
Comment (GiulioZizzo) [reply art/estimators/certification/deep_z/pytorch.py:58]: We are following the coding pattern in adversarial-robustness-toolbox/art/estimators/classification/pytorch.py which uses an ignore.

File: art/estimators/certification/deep_z/pytorch.py (hunk 7 above)
Comment (GiulioZizzo) [reply art/estimators/certification/deep_z/pytorch.py:141]: Ignored due to `nn.Conv2d` requiring the type hinting indicated in the init of `ZonoConv(torch.nn.Module)`. However, reading layer information using `module.X` can return types of a slightly different form (eg `Tuple[int, ...]` vs `Tuple[int]`) which nonetheless both function when passed into `nn.Conv2d`. Hence, using an ignore to suppress the warning.

**Review by beat-buesser — APPROVED** [review]
Hi @GiulioZizzo Thank you very much for implementing robustness certification based on zonotope domains!

**Pull request merged**